R/shiny-options.R

```diff
@@ -127,30 +149,76 @@ shinyOptions <- function(...) {
   newOpts <- list(...)
 
   if (length(newOpts) > 0) {
+    # If we're within a session, modify at the session level.
+    session <- getDefaultReactiveDomain()
+    if (!is.null(session)) {
+      # Modify session-level-options
+      session$options <- dropNulls(mergeVectors(session$options, newOpts))
+      return(invisible(session$options))
+    }
+
+    # If not in a session, but we have a currently running app, modify options
+    # at the app level.
+    app_state <- getCurrentAppState()
+    if (!is.null(app_state)) {
+      # Modify app-level options
+      app_state$options <- dropNulls(mergeVectors(app_state$options, newOpts))
+      return(invisible(app_state$options))
+    }
+
+    # If no currently running app, modify global options and return them.
     .globals$options <- dropNulls(mergeVectors(.globals$options, newOpts))
-    invisible(.globals$options)
-  } else {
-    .globals$options
+    return(invisible(.globals$options))
   }
-}
 
+  # If not setting any options, just return current option set, visibly.
 
-# Eval an expression with a new option set
-withLocalOptions <- function(expr) {
-  oldOptionSet <- .globals$options
-  on.exit(.globals$options <- oldOptionSet)
+  session <- getDefaultReactiveDomain()
+  if (!is.null(session)) {
+    return(session$options)
+  }
+
+  app_state <- getCurrentAppState()
+  if (!is.null(app_state)) {
+    return(app_state$options)
+  }
 
-  expr
+  return(.globals$options)
```

Comment:
That specific code isn't quite right, since a value could be cleared at a more specific scope. If someone does that, we don't want it to fall through to the value at a larger scope (if we did that, then if an option were set at the global scope, there would be no way to unset it in an app or session level scope.)

I thought about making a function for it, but the little details for each use case made it complicated enough that I decided against it.